File: prettier-plugin-csharpier/src/Printer/Types/FromClause.ts
import { PrintMethod } from "../PrintMethod";
import { HasIdentifier, SyntaxToken, printIdentifier, printSyntaxToken, SyntaxTreeNode } from "../SyntaxTreeNode";
import { concat, group, hardline, indent, join, softline, line, doubleHardline } from "../Builders";

export interface FromClauseNode extends SyntaxTreeNode<"FromClause"> {
    fromKeyword: SyntaxToken;
    type?: SyntaxTreeNode;
    identifier: SyntaxToken;
    inKeyword: SyntaxToken;
    expression: SyntaxTreeNode;
}

export const printFromClause: PrintMethod<FromClauseNode> = (path, options, print) => {
    const node = path.getValue();
    return concat([
        printSyntaxToken(node.fromKeyword),
        " ",
        printIdentifier(node),
        " ",
        printSyntaxToken(node.inKeyword),
        " ",
        path.call(print, "expression"),
    ]);
};

File: Src/CSharpier/SyntaxPrinter/SyntaxNodePrinters/FromClause.cs
namespace CSharpier.SyntaxPrinter.SyntaxNodePrinters;

internal static class FromClause
{
    public static Doc Print(FromClauseSyntax node, FormattingContext context)
    {
        return Doc.Concat(
            Token.PrintWithSuffix(node.FromKeyword, " ", context),
            node.Type != null ? Doc.Concat(Node.Print(node.Type, context), " ") : Doc.Null,
            Token.PrintWithSuffix(node.Identifier, " ", context),
            Token.PrintWithSuffix(node.InKeyword, " ", context),
            Node.Print(node.Expression, context)
        );
    }
}
